File: paleta/window.py
# window.py
#
# Copyright 2023 nate-xyz
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# SPDX-License-Identifier: GPL-3.0-or-later

from gi.repository import Adw, Gtk, Gdk, Gio

# from paleta.pages.image_drop import ImageDropPage
# from paleta.pages.palettes import PalettePage
from .pages import ImageDropPage, PalettePage
from .util import SUCCESS_GREEN, ERROR_RED

import os
import html

image_mime_types = ['image/jpeg', 'image/png', 'image/tiff', 'image/webp']

@Gtk.Template(resource_path='/io/github/nate_xyz/Paleta/window.ui')
class Window(Adw.ApplicationWindow):
    __gtype_name__ = 'Window'

    toast_overlay = Gtk.Template.Child(name="toast_overlay")
    header_bar = Gtk.Template.Child(name="header_bar")
    view_switcher_title = Gtk.Template.Child(name="view-switcher-title")
    stack = Gtk.Template.Child(name="stack")
    open_image_button = Gtk.Template.Child(name="open_image_button")
    image_drop_page = Gtk.Template.Child(name="image_drop_page")
    palette_page = Gtk.Template.Child(name="palette_page")
    edit_palette_button = Gtk.Template.Child(name="edit_palette_button")

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        app = kwargs['application']
        self.add_dialog()
        help_overlay = Gtk.Builder\
            .new_from_resource('/io/github/nate_xyz/Paleta/help-overlay.ui')\
            .get_object('help_overlay')
        self.set_help_overlay(help_overlay)
        
        self.setup_actions(app)

        self.db = app.db 
        self.db.window = self
        
        self.model = app.model
        
        self.saturate()
        
        if not self.db.try_loading_database():
            self.add_error_toast(_("Unable to load database."))
    
        self.clipboard = Gdk.Display.get_default().get_clipboard()
        self.setup_switcher_button()

        self.stack.connect('notify::visible-child-name', self.on_stack_switch)
        self.edit_palette_button.connect('clicked', lambda _button: self.palette_page.toggle_edit_mode())

    def setup_actions(self, app):
        app.set_accels_for_action("win.show-help-overlay", ['<Primary>question'])
        self.create_action(app, 'image-open', self.open_image_dialog.show, ['<Primary>o'])
        self.create_action(app, 'image-palette-extract', self.image_drop_page.thief_panel.start_extraction, ['<Primary>e', '<Primary>Return'])
        self.create_action(app, 'image-palette-save', self.image_drop_page.thief_panel.save_palette, ['<Primary>s'])
        self.create_action(app, 'image-page', self.go_to_image_drop_page, ['<Primary>i', '<Primary>d'])
        self.create_action(app, 'palette-edit-mode', self.palette_page.toggle_edit_mode, ['<Primary>t', '<Primary>m'])
        self.create_action(app, 'palette-add', self.palette_page.show_new_palette_dialog, ['<Primary>n','<Primary>a'])
        self.create_action(app, 'palette-page', self.go_to_palette_page, ['<Primary>p'])

    def on_stack_switch(self, stack, child_name):
        if stack.get_visible_child_name() == 'palette-stack-page' and len(self.model.get_palettes()) != 0:
            self.edit_palette_button.show()
        else:
            self.edit_palette_button.hide()
    
    def saturate(self):
        self.image_drop_page.saturate(self, self.db)
        self.palette_page.saturate(self, self.db, self.model)

    def setup_switcher_button(self):
        #adw switcher buttons
        squeezer = self.view_switcher_title.observe_children()[0]
        view_switcher = squeezer.observe_children()[0]
        self.switcher_buttons =  view_switcher.observe_children()
        #self.drop_button, self.palette_button = self.switcher_buttons
        for button in self.switcher_buttons:
            button.connect("clicked", self.switcher_button)

    def switcher_button(self, button):
        if self.check_switcher_title_bug(button):
            self.replace_switcher()

    def check_switcher_title_bug(self, active_button):
        error_string = 'button.flat.horizontal.toggle:active:dir(ltr)'
        for button in self.switcher_buttons:
            style_context = button.get_style_context()
            check_string = style_context.to_string(Gtk.StyleContextPrintFlags.SHOW_CHANGE).split(' ')[0]
            if button != active_button and error_string==check_string:
                return True
        return False
    
    def replace_switcher(self):        
        self.header_bar.remove(self.view_switcher_title)
        self.view_switcher_title = Adw.ViewSwitcherTitle()
        self.view_switcher_title.set_stack(self.stack)
        self.header_bar.set_title_widget(self.view_switcher_title)
        
        self.switcher_buttons =  self.view_switcher_title.observe_children()[0].observe_children()[0].observe_children()

        for button in self.switcher_buttons:
            button.connect("clicked", self.switcher_button)

    def add_dialog(self):
        self.open_image_dialog = Gtk.FileChooserNative.new(title=_("Select an Image File"), 
                                                        parent=self, 
                                                        action=Gtk.FileChooserAction.OPEN, 
                                                        accept_label=_("Open Image"))

        f = Gtk.FileFilter()
        f.set_name(_("Image files"))
        for m in image_mime_types:
            f.add_mime_type(m)

        self.open_image_dialog.connect("response", self.open_response)
        self.open_image_dialog.add_filter(f)
        self.open_image_button.connect("clicked", lambda _button: self.open_image_dialog.show())
        self.image_drop_page.open_image_button.connect("clicked", lambda _button: self.open_image_dialog.show())

    def open_response(self, dialog, response):
        if response == Gtk.ResponseType.ACCEPT:
            image_uri = dialog.get_file().get_path()
            if self.image_drop_page.load_image(image_uri):
                self.go_to_image_drop_page()
                self.open_image_toast(image_uri)
            else:
                self.error_image_toast(image_uri)

    def add_toast(self, title: str, timeout: int = 1):
        toast = Adw.Toast.new(html.escape(title))
        toast.set_timeout(timeout)
        self.toast_overlay.add_toast(toast)

    def add_toast_markup(self, title: str, timeout: int = 1):
        toast = Adw.Toast.new(title)
        toast.set_timeout(timeout)
        self.toast_overlay.add_toast(toast)

    def error_image_toast(self, uri):
        base_name = os.path.basename(uri)
        # Translators: Do not replace {}
        self.add_error_toast("Could not open image: {}".format(base_name), 3)

    def open_image_toast(self, uri):
        base_name = html.escape(os.path.basename(uri))
        # Translators: Do not replace {base_name}, {SUCCESS_GREEN}, or the span tags, only translate "Opened image:"
        self.add_toast_markup(f"<span foreground={SUCCESS_GREEN}>Opened image:</span>  {base_name}")

    def add_success_toast(self, verb: str, msg: str, timeout: int = 1):
        toast = Adw.Toast.new(f"<span foreground={SUCCESS_GREEN}>{verb}</span> {html.escape(msg)}")
        toast.set_timeout(timeout)
        self.toast_overlay.add_toast(toast)

    def add_error_toast(self, error: str, timeout: int = 1):
        # Translators: Only replace "Error!"
        toast = Adw.Toast.new(_(f"<span foreground={ERROR_RED}>Error!</span> {html.escape(error)}"))
        toast.set_timeout(timeout)
        self.toast_overlay.add_toast(toast)

    def add_color_toast(self, hex_name, palette_name):
        # Translators: Do not replace {hex_name}, only translate "Added color" and "to palette"
        self.add_toast_markup(_(f"Added color <span foreground=\"{hex_name}\">{hex_name}</span> to palette «{html.escape(palette_name)}»."))

    def remove_color_toast(self, hex_name, palette_name):
        # Translators: Do not replace {hex_name}, only translate "Removed color" and "from palette"
        self.add_toast_markup(_(f"Removed color <span foreground=\"{hex_name}\">{hex_name}</span> from palette «{html.escape(palette_name)}»."))

    def copy_color(self, hex_name):
        # Translators: Do not replace {hex_name}, only translate "Copied color" and "to clipboard!"
        self.add_toast_markup(_(f"Copied color <span foreground=\"{hex_name}\">{hex_name}</span> to clipboard!"))
        self.clipboard.set(hex_name)

    def go_to_image_drop_page(self):
        if self.stack.get_visible_child_name() != 'drop-stack-page':
            self.stack.set_visible_child(self.image_drop_page)

    def go_to_palette_page(self):
        if self.stack.get_visible_child_name() != 'palette-stack-page':
            self.stack.set_visible_child(self.palette_page)

    def create_action(self, app, name, callback=None, shortcuts=None):
        """Add a window action.

        Args:
            name: the name of the action
            callback: the function to be called when the action is
              activated
            shortcuts: an optional list of accelerators
        """
        action = Gio.SimpleAction.new(name, None)
        if callback != None:
            action.connect("activate", lambda _widget, _parameter: callback())
        self.add_action(action)
        if shortcuts:
            app.set_accels_for_action(f"win.{name}", shortcuts)

File: src/window.rs
/* window.rs
 *
 * SPDX-FileCopyrightText: 2023 nate-xyz
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

use adw::subclass::prelude::*;
use gtk::{prelude::*, gio, glib, glib::{clone, Sender}, gdk, CompositeTemplate};
use gtk_macros::send;

use std::{cell::RefCell, error::Error};
use log::{debug, error};

use crate::pages::{
    image_drop::image_drop_page::ImageDropPage,
    palette::palette_page::PalettePage,
};
use crate::database::DatabaseAction;
use crate::toasts::add_error_toast;
use crate::i18n::i18n;

use super::util::{database, model, settings_manager};

mod imp {
    use super::*;

    #[derive(Debug, CompositeTemplate)]
    #[template(resource = "/io/github/nate_xyz/Paleta/window.ui")]
    pub struct Window {

        #[template_child(id = "toast_overlay")]
        pub toast_overlay: TemplateChild<adw::ToastOverlay>,
        
        #[template_child(id = "header_bar")]
        pub header_bar: TemplateChild<adw::HeaderBar>,
        
        #[template_child(id = "view-switcher-title")]
        pub view_switcher_title: TemplateChild<adw::ViewSwitcherTitle>,

        #[template_child(id = "open_image_button")]
        pub open_image_button: TemplateChild<gtk::Button>,

        #[template_child(id = "stack")]
        pub stack: TemplateChild<adw::ViewStack>,

        #[template_child(id = "image_drop_page")]
        pub image_drop_page: TemplateChild<ImageDropPage>,

        #[template_child(id = "palette_page")]
        pub palette_page: TemplateChild<PalettePage>,

        #[template_child(id = "edit_palette_button")]
        pub edit_palette_button: TemplateChild<gtk::Button>,

        pub clipboard: Option<gdk::Clipboard>,
        pub open_image_dialog: RefCell<Option<gtk::FileChooserNative>>,
        pub settings: gio::Settings,
        pub db_sender: Sender<DatabaseAction>,
    }

    #[glib::object_subclass]
    impl ObjectSubclass for Window {
        const NAME: &'static str = "Window";
        type Type = super::Window;
        type ParentType = adw::ApplicationWindow;

        fn class_init(klass: &mut Self::Class) {
            klass.bind_template();
        }

        fn instance_init(obj: &glib::subclass::InitializingObject<Self>) {
            obj.init_template();
        }

        fn new() -> Self {
            Self {
                toast_overlay: TemplateChild::default(),
                header_bar: TemplateChild::default(),
                view_switcher_title: TemplateChild::default(),
                open_image_button: TemplateChild::default(),
                stack: TemplateChild::default(),
                image_drop_page: TemplateChild::default(),
                palette_page: TemplateChild::default(),
                edit_palette_button: TemplateChild::default(),
                clipboard: Some(gdk::Display::default().unwrap().clipboard()),
                open_image_dialog: RefCell::new(None),
                settings: settings_manager(),
                db_sender: database().sender(),
            }
        }

    }

    impl ObjectImpl for Window {}
    impl WidgetImpl for Window {}
    impl WindowImpl for Window {}
    impl ApplicationWindowImpl for Window {}
    impl AdwApplicationWindowImpl for Window {}
}

glib::wrapper! {
    pub struct Window(ObjectSubclass<imp::Window>)
        @extends gtk::Widget, gtk::Window, gtk::ApplicationWindow, adw::ApplicationWindow,
        @implements gio::ActionGroup, gio::ActionMap;
}

impl Window {
    pub fn new<P: glib::IsA<gtk::Application>>(application: &P) -> Self {
        let window: Window = glib::Object::builder()
            .property("application", application)
            .build();
        window.setup();
        window
    }


    fn setup(&self) {
        let imp = self.imp();

        self.setup_settings();
        self.bind_signals();
        self.add_dialog();

        send!(imp.db_sender, DatabaseAction::TryLoadingDataBase);
    }

    // fn add_help_overlay(&self) {
    //     let help_overlay = gtk::Builder::from_resource("/io/github/nate_xyz/Paleta/help-overlay.ui").object::<gtk::ShortcutsWindow>("help_overlay").unwrap();
    //     self.set_help_overlay(Some(&help_overlay));
    // }

    fn bind_signals(&self) {
        debug!("bind signals");
        let imp = self.imp();

        // self.edit_palette_button.connect('clicked', lambda _button: self.palette_page.toggle_edit_mode())
        imp.edit_palette_button.connect_clicked(
            clone!(@strong self as this => @default-panic, move |_button| {
                this.imp().palette_page.toggle_edit_mode();
            }),
        );

        // self.stack.connect('notify::visible-child-name', self.on_stack_switch)
        imp.stack.connect_notify_local(
            Some("visible-child-name"),
            clone!(@weak self as this => move |stack, _| {
                if stack.visible_child_name().unwrap().as_str() == "palette-stack-page" && model().palettes().len() != 0 {
                    this.imp().edit_palette_button.show();
                } else {
                    this.imp().edit_palette_button.hide();
                }
            }),
        );

        self.connect_local(
            "unrealize",
            false,
            clone!(@strong self as this => @default-return None, move |_value| {
                this.save_window_props();
                None
            }),
        );
    }

    fn go_to_image_drop_page(&self) {
        if self.imp().stack.visible_child_name().unwrap().as_str() != "drop-stack-page" {
            self.imp().stack.set_visible_child(&*self.imp().image_drop_page)
        }
    }

    pub fn go_to_palette_page(&self) {
        if self.imp().stack.visible_child_name().unwrap().as_str() != "palette-stack-page" {
            self.imp().stack.set_visible_child(&*self.imp().palette_page)
        }
    }

    pub fn add_toast(&self, toast: adw::Toast) {
        self.imp().toast_overlay.add_toast(toast);
    }

    fn add_dialog(&self) {
        let dialog = gtk::FileChooserNative::builder()
            .accept_label(&i18n("_Open Image"))
            .cancel_label(&i18n("_Cancel"))
            .modal(true)
            .title(&i18n("Select an Image File"))
            .action(gtk::FileChooserAction::Open)
            .select_multiple(false)
            .transient_for(self)
            .build();

        //let filter = gtk::FileFilter::new();
        // gtk::FileFilter::set_name(&filter, Some(&i18n("Image files")));
        // filter.add_mime_type("image/*");
        // dialog.add_filter(&filter);
        dialog.connect_response(
            clone!(@weak self as this => move |dialog, response| {
                if response == gtk::ResponseType::Accept {
                    match this.load_image(dialog.file()) {
                        Ok(uri) => {
                            if this.imp().image_drop_page.load_image(uri.as_str()) {
                                this.go_to_image_drop_page();
                            } else {
                                add_error_toast(i18n("Unable to load file."));
                            }
                        },
                        Err(e) => {
                            add_error_toast(i18n("Unable to load file."));
                            error!("{}", e);
                        },
                    }
                }
            }),
        );

        self.imp().open_image_dialog.replace(Some(dialog));

        self.imp().open_image_button.connect_clicked(
            clone!(@strong self as this => @default-panic, move |_button| {
                this.imp().open_image_dialog.borrow().as_ref().unwrap().show();
            }),
        );

        self.imp().image_drop_page.imp().open_image_button.connect_clicked(
            clone!(@strong self as this => @default-panic, move |_button| {
                this.imp().open_image_dialog.borrow().as_ref().unwrap().show();
            }),
        );

        
    }

    fn load_image(&self, file_option: Option<gio::File>) -> Result<String, Box<dyn Error>> {
        let file = file_option.ok_or("No file.")?;
        let path = file.path().ok_or("No path.")?;
        let uri = path.to_str().ok_or("No uri.")?.to_owned();
        Ok(uri)
    }

    pub fn edit_button_set_visible(&self, visible: bool) {
        let button: &gtk::Button = self.imp().edit_palette_button.as_ref();
        button.set_visible(visible);
    }

    pub fn edit_button_mode(&self, mode: bool) {
        let button: &gtk::Button = self.imp().edit_palette_button.as_ref();
        if mode {
            button.set_css_classes(&[&"opaque", &"edit-action-button"]);
        } else {
            button.set_css_classes(&[&"flat"]);
        }
    }

    pub fn copy_color(&self, hex_name: String) {
        self.imp().clipboard.as_ref().unwrap().set_text(hex_name.as_str());
    }

    /*
    GIO SETTINGS
    */

    fn setup_settings(&self) {
        let imp = self.imp();

        let width = imp.settings.int("window-width");
        let height = imp.settings.int("window-height");
        let maximized = imp.settings.boolean("window-maximized");

        self.set_default_size(width, height);
        self.set_maximized(maximized);
    }

    fn save_window_props(&self) {
        let imp = self.imp();

        let (width, height) = self.default_size();
        let maximized = self.is_maximized();

        _ = imp.settings.set_int("window-width", width);
        _ = imp.settings.set_int("window-height", height);
        _ = imp.settings.set_boolean("window-maximized", maximized);
    }
}
